src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1112 @@
+    return {
+      ...serializeTarget(sprite, {
+        stage,
+
+        getBroadcastId,
+
+        // Broadcasts are stored on the stage, not on any sprite.
+        broadcasts: {}
+      }),
+      isStage: false,
       x: sprite.x,
       y: sprite.y,
       size: sprite.size,
       direction: sprite.direction,
-      visible: sprite.visible,
+      rotationStyle: rotationStyleMap[sprite.rotationStyle],
       draggable: sprite.isDraggable,
-      isStage: false,
-      volume: sprite.volume
+      visible: sprite.visible
+    };
+  }
+
+  interface SerializeStageOptions {
+    getBroadcastId: GetBroadcastId;
+
+    broadcasts: sb3.Target["broadcasts"];
+    tempo: number;
+    textToSpeechLanguage: TextToSpeechLanguage;
+    videoState: "on" | "off";
+    videoTransparency: number;
+  }
+
+  function serializeStage(stage: Stage, options: SerializeStageOptions): sb3.Stage {
+    // Serialize a stage. Extending from a serialized target, the stage carries
+    // additional properties for values shared across the project - notably,
+    // the broadcast dictionary, as well as values for some extensions.
+
+    const { broadcasts, getBroadcastId } = options;
+    return {
+      ...serializeTarget(stage, { broadcasts, getBroadcastId, stage }),
+      isStage: true,
+      tempo: options.tempo,
+      textToSpeechLanguage: options.textToSpeechLanguage,
+      videoState: options.videoState,
+      videoTransparency: options.videoTransparency
+    };
+  }
+
+  function serializeProject(project: Project): sb3.ProjectJSON {
+    // Serialize a project. This is the master function used when project.toSb3
+    // is called. The main purpose of serializeProject is to serialize each
+    // target (sprite or stage) and collect them together in the final output
+    // format. It also provides utility functions shared across every target's
+    // serialization, e.g. broadcast utilities.
+
+    // Maintain broadcast dictionaries in both name <-> id directions. sb-edit
+    // operates on broadcast names, but Scratch 3.0 largely requires access to
+    // their IDs as well, so we generate IDs for names as they are required.
+    // Name -> ID is for converting sb-edit names and is the backbone behind
+    // getBroadcastId; ID -> name is the format which is directly serialized
+    // on the stage object. (Only the stage need carries a filled-out broadcast
+    // dictionary in Scratch 3.0.)
+    let broadcastNameToId: { [name: string]: string } = {};
+    let broadcastIdToName: { [id: string]: string } = {};
+
+    const getBroadcastId = (name: string): string => {
+      if (!(name in broadcastNameToId)) {
+        const id = generateId();
+        broadcastNameToId[name] = id;
+        broadcastIdToName[id] = name;
+      }
+      return broadcastNameToId[name];
     };
-    return sb3Sprite;
-  });
+
+    // Preemptively parse through all the broadcast inputs in the project and
+    // generate a broadcast ID for each of them.
```

Comment:
What's the reasoning behind doing this preemptively when it looks like `getBroadcastId` handles this correctly if doing this non-preemptively?

Reply:
We preemptively do so that we can find the alphabetically first broadcast name below - but you're right that getBroadcastId will handle it irregardless of if we parse preemptively, so I clarified the logic in 5144e3918d05e9832f4357f45bf23d5095924392.